src/libguac/mem.c

```diff
@@ -0,0 +1,237 @@
+        return 0;
+
+    /* Multiply all provided terms together */
+    ssize_t size = *(terms++);
+    while (--term_count) {
+
+        size_t term = *(terms++);
+
+        /* Fail if including this additional term would wrap past zero */
+        if (term > size)
+            return 1;
+
+        size -= term;
+
+    }
+
+    *result = size;
+    return 0;
+
+}
+
+size_t PRIV_guac_mem_ckd_mul_or_die(size_t factor_count, const size_t* factors) {
+
+    /* Perform request multiplication, aborting the entire process if the
+     * calculation overflows */
+    size_t result = 0;
+    if (PRIV_guac_mem_ckd_mul(&result, factor_count, factors))
+        abort();
+
+    return result;
+
+}
+
+size_t PRIV_guac_mem_ckd_add_or_die(size_t term_count, const size_t* terms) {
+
+    /* Perform request addition, aborting the entire process if the calculation
+     * overflows */
+    size_t result = 0;
+    if (PRIV_guac_mem_ckd_add(&result, term_count, terms))
+        abort();
+
+    return result;
+
+}
+
+size_t PRIV_guac_mem_ckd_sub_or_die(size_t term_count, const size_t* terms) {
+
+    /* Perform request subtraction, aborting the entire process if the
+     * calculation overflows */
+    size_t result = 0;
+    if (PRIV_guac_mem_ckd_sub(&result, term_count, terms))
+        abort();
+
+    return result;
+
+}
+
+void* PRIV_guac_mem_alloc(size_t factor_count, const size_t* factors) {
+
+    size_t size = 0;
+
+    if (PRIV_guac_mem_ckd_mul(&size, factor_count, factors)) {
+        guac_error = GUAC_STATUS_NO_MEMORY;
+        return NULL;
+    }
+    else if (size == 0)
+        return NULL;
+
+    void* mem = malloc(size);
+    if (mem == NULL) {
+        /* C does not require that malloc() set errno (though POSIX does). For
+         * portability, we set guac_error here regardless of the underlying
+         * behavior of malloc(). */
+        guac_error = GUAC_STATUS_NO_MEMORY;
+    }
+
+    return mem;
+
+}
+
+void* PRIV_guac_mem_zalloc(size_t factor_count, const size_t* factors) {
+
+    size_t size = 0;
+
+    if (PRIV_guac_mem_ckd_mul(&size, factor_count, factors)) {
+        guac_error = GUAC_STATUS_NO_MEMORY;
+        return NULL;
+    }
+    else if (size == 0)
+        return NULL;
+
+    void* mem = calloc(1, size);
+    if (mem == NULL) {
+        /* C does not require that calloc() set errno (though POSIX does). For
+         * portability, we set guac_error here regardless of the underlying
+         * behavior of calloc(). */
+        guac_error = GUAC_STATUS_NO_MEMORY;
+    }
+
+    return mem;
+
+}
+
+void* PRIV_guac_realloc(void* mem, size_t factor_count, const size_t* factors) {
+
+    size_t size = 0;
+
+    if (PRIV_guac_mem_ckd_mul(&size, factor_count, factors)) {
+        guac_error = GUAC_STATUS_NO_MEMORY;
+        return NULL;
+    }
+
+    /* Resize to 0 is equivalent to free() */
+    if (size == 0) {
+        guac_mem_free(mem);
+        return NULL;
+    }
+
+    void* resized_mem = realloc(mem, size);
```

Comment:
Should we free the mem if the realloc fails?

Reply:
No, it's on the caller to decide whether to keep the original memory if it's not possible to reallocate. This is necessary to ensure the caller can still decide to take some other action in the case of allocation failure. We can't deviate from that without breaking compatibility with the core expectations of a call to `realloc()`.

I think a `guac_mem_realloc_or_die()` could make sense, though.

Reply:
that's a solid point and I like the idea of `guac_mem_realloc_or_die` to go with the rest of the `or_die` 💀 functions

Reply:
Added via rebase, along with unit tests for both variants.